src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -214,26 +214,57 @@ export class WorkspaceSavedObjectsClientWrapper {
         );
         if (options.workspaces) {
           const isEveryWorkspaceIsPermitted = options.workspaces.every((item) =>
-            // TODO modify this line to use permittedWorkspaceIds if public workspace is also a workspace
-            ['public', ...(permittedWorkspaceIds || [])]?.includes(item)
+            (permittedWorkspaceIds || []).includes(item)
           );
           if (!isEveryWorkspaceIsPermitted) {
             throw generateWorkspacePermissionError();
           }
         } else {
           const queryDSL = ACL.genereateGetPermittedSavedObjectsQueryDSL(
-            [
-              PermissionMode.LibraryRead,
-              PermissionMode.LibraryWrite,
-              PermissionMode.Management,
-              PermissionMode.Read,
-              PermissionMode.Write,
-            ],
+            [PermissionMode.Read, PermissionMode.Write],
             principals,
             options.type
           );
-          options.workspaces = permittedWorkspaceIds;
-          options.queryDSL = queryDSL;
+          options.workspaces = undefined;
+          /**
+           * Select all the docs that
+           * 1. ACL matches right or write permission OR
+           * 2. workspaces matches library_read or library_write or management OR
+           * 3. Records without workspaces field (Advances settings)
+           */
+          options.queryDSL = {
```

Comment:
@gaobinlong Please help to review this query DSL, I do not know if it is easy to read and with good search performance.